File: lib/features/meals/domain/food_nutrition_math.dart
import 'models.dart';
import 'food_serving_kind.dart';
import 'package:wellness_app/l10n/app_localizations.dart';
import 'package:wellness_app/services/language_service.dart';

class NutritionMacros {
  const NutritionMacros({
    required this.kcal,
    required this.protein,
    required this.carbs,
    required this.fat,
  });

  final double kcal;
  final double protein;
  final double carbs;
  final double fat;

  Map<String, double> toMap() => {
        'kcal': kcal,
        'protein': protein,
        'carbs': carbs,
        'fat': fat,
      };
}

extension FoodItemServing on FoodItem {
  FoodServingKind get servingKind => FoodServingKindParser.fromLegacyUnit(unit);
}

/// Single source of truth for display ↔ stored quantity and macro math.
class FoodNutritionMath {
  FoodNutritionMath._();

  static FoodServingKind kindFor(FoodItem food) => food.servingKind;

  static double macroMultiplier(FoodServingKind kind, double storedQuantity) {
    return storedQuantity;
  }

  static double displayQuantity(FoodItem food, double storedQuantity) {
    switch (food.servingKind) {
      case FoodServingKind.per100g:
        return storedQuantity * 100;
      case FoodServingKind.perGram:
      case FoodServingKind.perMl:
      case FoodServingKind.perOz:
      case FoodServingKind.perCount:
        return storedQuantity;
    }
  }

  static double storedQuantity(FoodItem food, double displayQuantity) {
    switch (food.servingKind) {
      case FoodServingKind.per100g:
        return displayQuantity / 100;
      case FoodServingKind.perGram:
      case FoodServingKind.perMl:
      case FoodServingKind.perOz:
      case FoodServingKind.perCount:
        return displayQuantity;
    }
  }

  /// Localised form of a stored unit string.
  ///
  /// `FoodItem.unit` holds display tokens like "100g", "piece", "tbsp". They
  /// are part of the seeded data rather than the UI, so there is no ARB key
  /// per food -- one token map covers the whole catalog, the same way
  /// `FoodCategory.label` handles categories. A numeric prefix is kept and only
  /// its suffix translated, so "100g" becomes "100 ג" rather than being lost.
  static String localizedUnit(AppLanguage language, String unit) {
    if (language != AppLanguage.hebrew) return unit;
    const tokens = {
      'g': 'ג',
      'ml': 'מ״ל',
      'oz': 'אונקיה',
      'piece': 'יחידה',
      'pieces': 'יחידות',
      'slice': 'פרוסה',
      'tbsp': 'כף',
      'tsp': 'כפית',
      'cup': 'כוס',
      'scoop': 'סקופ',
      'serving': 'מנה',
      'can': 'פחית',
      'bottle': 'בקבוק',
    };
    final trimmed = unit.trim();
    if (tokens.containsKey(trimmed)) return tokens[trimmed]!;
    // "100g" / "300ml" -- keep the number, translate the suffix.
    final m = RegExp(r'^(\d+(?:\.\d+)?)\s*([a-zA-Z]+)$').firstMatch(trimmed);
    if (m != null) {
      final suffix = tokens[m.group(2)!.toLowerCase()];
      if (suffix != null) return '${m.group(1)} $suffix';
    }
    return unit;
  }

  static String displayUnitLabel(FoodItem food) {
    switch (food.servingKind) {
      case FoodServingKind.per100g:
        return 'g';
      case FoodServingKind.perGram:
        return 'g';
      case FoodServingKind.perMl:
        return 'ml';
      case FoodServingKind.perOz:
        return 'oz';
      case FoodServingKind.perCount:
        return food.unit;
    }
  }

  static NutritionMacros computeMacros(FoodItem food, double storedQuantity) {
    final multiplier = macroMultiplier(food.servingKind, storedQuantity);
    return NutritionMacros(
      kcal: food.kcalPerUnit * multiplier,
      protein: food.proteinPerUnit * multiplier,
      carbs: food.carbsPerUnit * multiplier,
      fat: food.fatPerUnit * multiplier,
    );
  }

  static NutritionMacros computeMacrosFromDisplay(
    FoodItem food,
    double displayQuantity,
  ) {
    return computeMacros(food, storedQuantity(food, displayQuantity));
  }

  static String formatAmountLine(FoodItem food, double storedQuantity) {
    final display = displayQuantity(food, storedQuantity);
    final unit = displayUnitLabel(food);
    return '${_formatNumber(display)} $unit';
  }

  /// Takes [l10n] rather than a BuildContext: this is domain code, and the
  /// caller already has the localisations in hand.
  static String explainUnit(AppLocalizations l10n, FoodItem food) {
    switch (food.servingKind) {
      case FoodServingKind.per100g:
        return l10n.unitPer100g;
      case FoodServingKind.perGram:
        return l10n.unitPerGram;
      case FoodServingKind.perMl:
        return l10n.unitPerMl;
      case FoodServingKind.perOz:
        return l10n.unitPerOz;
      case FoodServingKind.perCount:
        return l10n.unitPerCount(food.unit);
    }
  }

  static String _formatNumber(double value) {
    if (value == value.roundToDouble()) {
      return value.toInt().toString();
    }
    return value.toStringAsFixed(1);
  }
}

File: Packages/WellnessKit/Sources/WellnessDomain/FoodNutritionMath.swift
import Foundation
import WellnessModels

public enum FoodServingKindParser: Sendable {
    public static func fromLegacyUnit(_ unit: String) -> FoodServingKind {
        let u = unit.lowercased().trimmingCharacters(in: .whitespaces)

        if u == "100g" || u == "100 g" { return .per100g }
        if u == "g" || u == "gram" || u == "grams" { return .perGram }
        if u == "ml" || u == "milliliter" || u == "milliliters" { return .perMl }
        if u == "oz" || u == "ounce" || u == "ounces" { return .perOz }
        if ["piece", "slice", "tbsp", "scoop", "serving", "item", "each"].contains(u) {
            return .perCount
        }
        if u.range(of: #"^\d+g$"#, options: .regularExpression) != nil { return .perCount }
        return .perCount
    }

    public static func normalizeToCatalogUnit(_ unit: String) -> String {
        let kind = fromLegacyUnit(unit)
        if kind == .perCount, let u = FoodServingUnit(rawValue: unit), u.kind == .perCount {
            return unit
        }
        return FoodServingUnit.forKind(kind, legacy: unit).rawValue
    }
}

extension FoodItem {
    public var servingKind: FoodServingKind {
        FoodServingKindParser.fromLegacyUnit(unit)
    }
}

public enum FoodNutritionMath: Sendable {
    public static func kindFor(_ food: FoodItem) -> FoodServingKind {
        food.servingKind
    }

    public static func macroMultiplier(_ kind: FoodServingKind, storedQuantity: Double) -> Double {
        storedQuantity
    }

    public static func displayQuantity(_ food: FoodItem, storedQuantity: Double) -> Double {
        switch food.servingKind {
        case .per100g: storedQuantity * 100
        case .perGram, .perMl, .perOz, .perCount: storedQuantity
        }
    }

    public static func storedQuantity(_ food: FoodItem, displayQuantity: Double) -> Double {
        switch food.servingKind {
        case .per100g: displayQuantity / 100
        case .perGram, .perMl, .perOz, .perCount: displayQuantity
        }
    }

    public static func displayUnitLabel(_ food: FoodItem) -> String {
        switch food.servingKind {
        case .per100g, .perGram: "g"
        case .perMl: "ml"
        case .perOz: "oz"
        case .perCount: food.unit
        }
    }

    public static func computeMacros(_ food: FoodItem, storedQuantity: Double) -> NutritionMacros {
        let m = macroMultiplier(food.servingKind, storedQuantity: storedQuantity)
        return NutritionMacros(
            kcal: food.kcalPerUnit * m,
            protein: food.proteinPerUnit * m,
            carbs: food.carbsPerUnit * m,
            fat: food.fatPerUnit * m
        )
    }

    public static func computeMacrosFromDisplay(
        _ food: FoodItem,
        displayQuantity: Double
    ) -> NutritionMacros {
        computeMacros(food, storedQuantity: storedQuantity(food, displayQuantity: displayQuantity))
    }

    public static func formatAmountLine(_ food: FoodItem, storedQuantity: Double) -> String {
        let display = displayQuantity(food, storedQuantity: storedQuantity)
        let unit = displayUnitLabel(food)
        return "\(formatNumber(display)) \(unit)"
    }

    public static func localizedUnit(language: AppLanguage, unit: String) -> String {
        guard language == .hebrew else { return unit }
        let tokens: [String: String] = [
            "g": "ג", "ml": "מ״ל", "oz": "אונקיה",
            "piece": "יחידה", "pieces": "יחידות", "slice": "פרוסה",
            "tbsp": "כף", "tsp": "כפית", "cup": "כוס",
            "scoop": "סקופ", "serving": "מנה", "can": "פחית", "bottle": "בקבוק",
        ]
        let trimmed = unit.trimmingCharacters(in: .whitespaces)
        if let t = tokens[trimmed] { return t }
        if let range = trimmed.range(of: #"^(\d+(?:\.\d+)?)\s*([a-zA-Z]+)$"#, options: .regularExpression) {
            let matched = String(trimmed[range])
            let numEnd = matched.firstIndex(where: { $0.isLetter }) ?? matched.endIndex
            let num = matched[matched.startIndex..<numEnd].trimmingCharacters(in: .whitespaces)
            let alpha = matched[numEnd...].trimmingCharacters(in: .whitespaces).lowercased()
            if let suffix = tokens[alpha] {
                return "\(num) \(suffix)"
            }
        }
        return unit
    }

    private static func formatNumber(_ value: Double) -> String {
        if value == value.rounded(.towardZero) && value == Double(Int(value)) {
            return String(Int(value))
        }
        return String(format: "%.1f", value)
    }
}
